### PR TITLE
[model] added support for google Gemma-2b model

### PR DESCRIPTION
**Pull request by 936187425**
Integrating Gemma-2b

### CODE REVIEW

File: src/server/simple.cpp [hunk 1]
```diff
@@ -22,6 +22,8 @@ DEFINE_string(model_allow_patterns,
               "*.json,*.tiktoken,*.model,*.safetensors",
               "Allow patterns for model files.");
 
+DEFINE_string(token, "", "hugging face token");
```
Comment (guocuimi) [inline src/server/simple.cpp:25]: actually, you don't need pass token in this way. a better way is set a environment variable and huggingface_hub will load it automatically. `HUGGING_FACE_HUB_TOKEN`

**Review by guocuimi — COMMENTED** [review]
thanks for adding new model. let me know once your code is ready to review.

File: src/server/simple.cpp (hunk 1 above)
Comment (936187425) [reply src/server/simple.cpp:25]: understand！

**Comment by 936187425** [comment]
> thanks for adding new model. let me know once your code is ready to review.

You can review the code.  If there are any issues during the review process, please kindly inform me. thank you！

File: src/server/simple.cpp [hunk 2]
```diff
@@ -110,11 +110,11 @@ int main(int argc, char* argv[]) {
   stopping_criteria.eos_token_id = model_args.eos_token_id();
   stopping_criteria.stop_token_ids = model_args.stop_token_ids();
   stopping_criteria.max_context_length = model_args.max_position_embeddings();
-
   std::string prompt = "Enter a prompt: ";
-  std::cout << prompt;
-
+  // std::cout << prompt;
```
Comment (liutongxuan) [inline src/server/simple.cpp:114]: 可以把注释这个std::cout删掉

File: src/models/huggingface/gemma.h [hunk 3]
```diff
@@ -0,0 +1,458 @@
...
+  // tokens: [num_tokens]
+  // positions: [num_tokens] token pos in the sequence
+  // return: [num_tokens,hidden_size]
+  torch::Tensor forward(const torch::Tensor& tokens,
+                        const torch::Tensor& positions,
+                        std::vector<KVCache>& kv_caches,
+                        const InputParameters& input_params) {
+    auto h = model_(tokens, positions, kv_caches, input_params);
+
+    // LOG(INFO) << "GemmaForCausalLMImpl forward:" << h.sizes();
```
Comment (liutongxuan) [inline src/models/huggingface/gemma.h:356]: 注释的这两行可以删掉

File: src/models/huggingface/gemma.h [hunk 4]
```diff
@@ -0,0 +1,458 @@
...
+#include <torch/torch.h>
+
+#include "layers/attention/handler.h"
+#include "layers/embedding.h"
+#include "layers/normalization.h"
+// gemma model compatible with huggingface weight
+namespace llm::hf {
+
+// TODO only support the gemma-2B now
+enum GemmaType {
```
Comment (liutongxuan) [inline src/models/huggingface/gemma.h:12]: 这个gemmaType后面没有使用？

Comment (936187425) [reply src/models/huggingface/gemma.h:12]: > 这个gemmaType后面没有使用？

暂时仅适配gemma-2b

File: src/models/huggingface/gemma.h [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+                        std::vector<KVCache>& kv_caches,
+                        const InputParameters& input_params) {
+    // embedding tokens
+    auto h = embed_tokens_(tokens);
+
+    // normalize the embedding by sqrt(hidden_size)
+    h *= sqrt(modelArgs_.hidden_size());
+
+    // https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/gemma.py
+    // line:236
```
Comment (liutongxuan) [inline src/models/huggingface/gemma.h:285]: 这个注释需要吗？

File: src/models/huggingface/gemma.h [hunk 6]
```diff
@@ -0,0 +1,461 @@
...
+    StateDict new_state_dict =
+        state_dict.add_scalar_with_suffix("norm.weight", 1);
+
+    embed_tokens_->load_state_dict(new_state_dict.select("embed_tokens."));
+    // call each layer's load_state_dict function
+    for (int i = 0; i < layers_.size(); i++) {
+      layers_[i]->load_state_dict(
+          new_state_dict.select("layers." + std::to_string(i) + "."));
+    }
+    norm_->load_state_dict(new_state_dict.select("norm."));
```
Comment (guocuimi) [inline src/models/huggingface/gemma.h:303]: let's use select_with_transform isntead.
```
norm_->load_state_dict((state_dict.select_with_transform(
        "norm.", [](torch::Tensor tensor) { return tensor + 1.0f; }));
```

File: src/model_loader/state_dict.cpp [hunk 7]
```diff
@@ -238,6 +238,20 @@ StateDict StateDict::select(const std::string_view& prefix) const {
   }
   return {std::move(selected), shard_id_, num_shards_};
 }
+// add the scalar value into the selected tensors ending with suffix
+StateDict StateDict::add_scalar_with_suffix(const std::string_view& suffix,
```
Comment (guocuimi) [inline src/model_loader/state_dict.cpp:242]: we don't need to introduce a new function, let's use `select_with_transform` instead.

**Review by guocuimi — COMMENTED** [review]
thanks for adding Gemma. please help addressing comments.

**Review by guocuimi — APPROVED** [review]

**Pull request merged**